modeling/algorithm/local-weighted-slice-algorithm.go

```diff
@@ -0,0 +1,202 @@
+	if region.ZoneDetails == nil {
+		return nil, fmt.Errorf("zoneDetail should not be nil")
+	}
+	sliceGroups := map[string]types.EndpointSliceGroup{}
+	// endpointsAvailable stores zones with int number of endpoints available
+	endpointsAvailable := endpointsList{}
+	// endpointsNeeded stores zones with int number of endpoints needed
+	endpointsNeeded := endpointsList{}
+	// weightedEndpointsAvailable stores zones with float number of endpoints
+	// available
+	weightedEndpointsAvailable := endpointsList{}
+	// weightedEndpointsNeeded stores zones with float number of endpoints
+	// needed
+	weightedEndpointsNeeded := endpointsList{}
+	for zoneName, zone := range region.ZoneDetails {
+		var localGroup types.EndpointSliceGroup
+		localGroup.Label = zoneName
+		// this local sliceGroup should only receive traffic from current zone,
+		// this map tracks weights of traffic from different zones to this
+		// sliceGroup
+		localGroup.ZoneTrafficWeights = map[string]float64{zoneName: 1.0}
+		// this map stores composition of this sliceGroup, zoneName - number of
+		// endpoints from zoneName
+		localGroup.Composition = map[string]types.WeightedEndpoints{}
+
+		// calculate expected number of endpoints based on the proportion of
+		// nodes in this zone
+		expectedEndpoints := zone.NodesRatio * float64(region.TotalEndpoints)
+		// deviation: a negative value means need more endpoints from other
+		// sliceGroups, a positive value means need give out endpoints to other
+		// sliceGroups
+		deviation := float64(zone.Endpoints) - expectedEndpoints
+		// intDeviation is dealt same way as original local-slice algorithm that
+		// directly 'borrows' or 'sends' endpoints among zones
+		intDeviation := int(deviation)
+		if intDeviation == 0 {
+			localGroup.Composition[zoneName] = types.WeightedEndpoints{Number: int(expectedEndpoints), Weight: 1}
+		} else if intDeviation > 0 {
+			endpointsAvailable.push(endpointDeviation{name: zoneName, deviation: intDeviation})
+			localGroup.Composition[zoneName] = types.WeightedEndpoints{Number: int(expectedEndpoints), Weight: 1}
+		} else {
+			endpointsNeeded.push(endpointDeviation{name: zoneName, deviation: -intDeviation})
+			localGroup.Composition[zoneName] = types.WeightedEndpoints{Number: zone.Endpoints, Weight: 1}
+		}
+		sliceGroups[zoneName] = localGroup
+
+		// push 0.xx deviation into corresponding lists
+		floatDeviation := deviation - float64(intDeviation)
+		if floatDeviation > 0 {
+			// Intentionally set deviation to default 0, this means the zone
+			// itself needs float part from this endpoint otherwise means the
+			// endpoints will be shared by other zones only
+			weightedEndpointsAvailable.push(endpointDeviation{name: zoneName, weight: floatDeviation})
+		} else if floatDeviation < 0 {
+			weightedEndpointsNeeded.push(endpointDeviation{name: zoneName, deviation: 1, weight: -floatDeviation})
```

Comment:
Why is deviation 1 here? Why is weight -floatDeviation?

Reply:
Done